src/psyclone/psyGen.py

```diff
@@ -2771,19 +2740,65 @@ def gen_code(self, parent):
 
         :param parent: The parent of this kernel call in the f2pygen AST.
         :type parent: :py:calls:`psyclone.f2pygen.LoopGen`
+
+        :raises NotImplementedError: if there is a name clash preventing to \
+            module-inline the kernel without changing its name.
         '''
-        from psyclone.f2pygen import CallGen, UseGen
+        from psyclone.f2pygen import CallGen, UseGen, PSyIRGen
 
         # If the kernel has been transformed then we rename it. If it
         # is *not* being module inlined then we also write it to file.
         self.rename_and_write()
 
-        parent.add(CallGen(parent, self._name,
-                           self.arguments.raw_arg_list()))
+        # Add the subroutine call with the necessary arguments
+        arguments = self.arguments.raw_arg_list()
+        parent.add(CallGen(parent, self._name, arguments))
 
+        # Also add the subroutine declaration, this can just be the import
+        # statement, or the whole subroutine inlined into the module.
         if not self.module_inline:
             parent.add(UseGen(parent, name=self._module_name, only=True,
                               funcnames=[self._name]))
+        else:
+            # Find the root f2pygen module
```

Comment:
Does the f2pygen ast not have a `root` method? If not, then don't worry - there's no point adding one.

Reply:
Worse than that, it does have a `root` method but it doesn't return the ancestor of the `.parent` chain.

I think it uses "root" to mean the source of the content. Anyways it returns a fparser node, while in this case I want the root f2pygen node, hence I have to do it manually.